src/peergos/shared/user/fs/FileWrapper.java

```diff
@@ -694,29 +694,78 @@ public static int getNumberOfChunks(long size) {
-                                        return chunks.upload(network, parentLocation.owner, signer)
-                                                .thenCompose(fileLocation -> {
-                                                    WritableAbsoluteCapability fileWriteCap =
-                                                            new WritableAbsoluteCapability(owner(), signer.publicKeyHash,
-                                                                    locations.get(0).getMapKey(), fileKey, fileWriteKey);
-                                                    return addChildPointer(filename, fileWriteCap, network, random, 2);
-                                                });
-                                    }))
-                            )
-                    );
+            
+            return calculateMimeType(fileData, endIndex)
+                .thenCompose(mimeType -> fileData
+                        .reset()
+                        .thenCompose(resetReader -> {
+
+                                FileProperties fileProps = new FileProperties(filename, mimeType, endIndex,
+                                        LocalDateTime.now(), isHidden, Optional.empty()); 
+
+                                FileUploader chunks = new FileUploader(filename, mimeType, resetReader,
+                                        startIndex, endIndex, fileKey, parentLocation, dirParentKey, monitor, fileProps,
+                                        fragmenter, locations);
+
+                                SigningPrivateKeyAndPublicHash signer = signingPair();
+
+                                return chunks.upload(network, parentLocation.owner, signer)
+                                    .thenCompose(fileLocation -> {
+
+                                        WritableAbsoluteCapability fileWriteCap = new
+                                            WritableAbsoluteCapability(owner(),
+                                                    signer.publicKeyHash,
+                                                    locations.get(0).getMapKey(), fileKey,
+                                                    fileWriteKey);
+
+                                        return addChildPointer(filename, fileWriteCap, network, random, 2)
+                                            .thenCompose(pointer -> fileData
+                                                    .reset()
+                                                    .thenCompose(resetAgain -> { 
+                                                        return generateThumbnailAndUpdate(pointer, filename, resetAgain, network, thumbnailSrcImageSize, isHidden, mimeType, endIndex, LocalDateTime.now());
+                                            }));
+
+                                    });
+                        })
+            );
+
         });
+
     }
 
+    private CompletableFuture<FileWrapper> generateThumbnailAndUpdate(
+            FileWrapper hashPointer,
```

Comment:
parent is probably a clearer name than hashPointer

Reply:
Agreed 👍